### PR TITLE
#37-New/add edit opportunities

### PR DESCRIPTION
**Pull request by deimonbody**
Page opportunities version 1.0

### CODE REVIEW

**Review by AndriyCh1 — APPROVED** [review]

File: backend/src/api/routes/index.ts [hunk 1]
```diff
@@ -4,12 +4,13 @@ import { errorHandlerMiddleware } from '../middlewares/error-handler-middleware'
...
 import userRoute from './user-route';
 import tagsRoute from './tags-route';
+import opportunitiesRoute from './opportunities-route';
 
 const routes = (app: Express): void => {
   app.use('/api/auth', authenticationRoute);
   app.use('/api/user', userRoute);
   app.use('/api/company/tags', tagsRoute);
-
+  app.use('/company/opportunities', opportunitiesRoute);
```
Comment (aleksandr-kuharenko) [inline backend/src/api/routes/index.ts]:
```suggestion
  app.use('/api/company/opportunities', opportunitiesRoute);
```

File: backend/src/api/routes/opportunities-route.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import { Router } from 'express';
+import { run } from '~/common/helpers/route.helper';
+import { getOpportunities } from '~/services/opportunity.service';
+
+const router: Router = Router();
+router.get('/',run((_)=>getOpportunities()));
+export default router;
```
Comment (aleksandr-kuharenko) [inline backend/src/api/routes/opportunities-route.ts]: please format this file

File: backend/src/data/seed-data/opp_tag.data.ts [hunk 3]
```diff
@@ -0,0 +1 @@
+export const opp_tags = [{},{},{}];
```
Comment (aleksandr-kuharenko) [inline backend/src/data/seed-data/opp_tag.data.ts]: is it still needed?

File: backend/src/data/seeders/index.ts [hunk 4]
```diff
@@ -4,9 +4,16 @@ import ormconfig from '~/config/ormconfig';
...
-});
+async function Connection():Promise<void>{
+  await createConnection(ormconfig);
+  await CompanySeeder.execute();
+  await UserSeeder.execute();
+  await OpportunitySeeder.execute();
+  await TagsSeeder.execute();
+  await OpportunityTagSeeder.execute();
+}
+Connection();
```
Comment (aleksandr-kuharenko) [inline backend/src/data/seeders/index.ts]: add try...catch block and log error message

File: frontend/package.json [hunk 5]
```diff
@@ -31,7 +32,8 @@
     "react-notifications": "1.7.2",
     "react-redux": "7.2.6",
     "react-router-dom": "6.2.1",
-    "react-scripts": "5.0.0"
+    "react-scripts": "5.0.0",
+    "uuid": "^8.3.2"
```
Comment (aleksandr-kuharenko) [inline frontend/package.json]: remove this lib if it is not used anywhere

File: frontend/package.json [hunk 6]
```diff
@@ -17,6 +17,7 @@
     "@hookform/error-message": "2.0.0",
     "@hookform/resolvers": "2.8.5",
     "@reduxjs/toolkit": "1.7.1",
+    "@types/uuid": "^8.3.4",
```
Comment (aleksandr-kuharenko) [inline frontend/package.json]: same

**Review by aleksandr-kuharenko — CHANGES_REQUESTED** [review]

File: frontend/src/components/opportunities-page/OpportunityModal.tsx [hunk 7]
```diff
@@ -0,0 +1,52 @@
+import { useAppDispatch } from 'hooks/hooks';
+import { useRef } from 'react';
+import { X } from 'react-bootstrap-icons';
+// import { fetchAddNewOpp } from 'store/opportunities/actions';
```
Comment (nikita-remeslov) [inline frontend/src/components/opportunities-page/OpportunityModal.tsx]: Do you need these comments?

File: frontend/src/components/opportunities-page/styles.scss [hunk 8]
```diff
@@ -0,0 +1,50 @@
+.opportunities{
+    // &__block{
+    //     cursor: pointer;
```
Comment (nikita-remeslov) [inline frontend/src/components/opportunities-page/styles.scss]: The same

File: frontend/src/store/store.ts [hunk 9]
```diff
@@ -19,3 +19,6 @@ const store = configureStore({
 });
 
 export { store };
+export type RootState = ReturnType<typeof store.getState>;
+// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
```
Comment (nikita-remeslov) [inline frontend/src/store/store.ts]: the same

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]

**Comment by nikita-remeslov** [comment]
Fix comment for this PR

**Comment by nikita-remeslov** [comment]
Fix conflicts

File: frontend/src/components/main-page/opportunities/opportunities.scss [hunk 10]
```diff
@@ -1,10 +1,10 @@
 .opportunities {
   &__cart {
     cursor: pointer;
-    transition: all 0.5s linear;
-    &:hover {
-      background-color: rgba(207, 204, 204, 0.781) !important;
-    }
+    // transition: all 0.5s linear;
```
Comment (nikita-remeslov) [inline frontend/src/components/main-page/opportunities/opportunities.scss]: Remove it if you don't need this

File: frontend/src/components/main-page/opportunities/opportunityList.tsx [hunk 11]
```diff
@@ -0,0 +1,54 @@
+import OpportunityItem from './opportunityItem';
+import './opportunities.scss';
+const OpportunityList: React.FC = () => {
+  const opportunities = [
+    {
```
Comment (nikita-remeslov) [inline frontend/src/components/main-page/opportunities/opportunityList.tsx]: Move mock to redux

File: frontend/src/components/opportunities-page/Follow.tsx [hunk 12]
```diff
@@ -0,0 +1,33 @@
+import { Bookmark, BookmarkFill } from 'react-bootstrap-icons';
+import { IOpportunity } from 'store/opportunities/common';
+import { useAppDispatch } from 'hooks/hooks';
+import * as opportunitiesActions from '../../store/opportunities/actions';
+
+interface Props extends IOpportunity {}
```
Comment (nikita-remeslov) [inline frontend/src/components/opportunities-page/Follow.tsx:6]: I think it should work like this `interface Props extends IOpportunity`

File: frontend/src/components/opportunities-page/opportunitiesPageItem.tsx [hunk 13]
```diff
@@ -0,0 +1,30 @@
+import Follow from './Follow';
+import OpportunityItem from 'components/main-page/opportunities/opportunityItem';
+import { IOpportunity } from 'store/opportunities/common';
+
+interface Props extends IOpportunity {}
```
Comment (nikita-remeslov) [inline frontend/src/components/opportunities-page/opportunitiesPageItem.tsx:5]: same

File: frontend/src/components/profile/components/profile-info/profile-info.scss [hunk 14]
```diff
@@ -1,4 +1,4 @@
-@import '~assets/scss/utils';
+@import 'src/assets/scss/styles';
```
Comment (nikita-remeslov) [inline frontend/src/components/profile/components/profile-info/profile-info.scss]: check my comments about styles in our team chat. Also if you use imports, please use alias if needed

File: frontend/src/components/profile/components/profile-info/tabs/tabs.scss [hunk 15]
```diff
@@ -1,4 +1,4 @@
-@import '~assets/scss/utils';
+@import 'src/assets/scss/styles';
```
Comment (nikita-remeslov) [inline frontend/src/components/profile/components/profile-info/tabs/tabs.scss]: the same, remove this changes

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]

File: backend/src/api/routes/opportunities-route.ts [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+import { run } from '~/common/helpers/route.helper';
+import {
+  getOpportunities,
+  createOpportunities,
+} from '~/services/opportunity.service';
+
+const router: Router = Router();
+router.get(
+  '/',
+  run((_) => getOpportunities()),
```
Comment (aleksandr-kuharenko) [inline backend/src/api/routes/opportunities-route.ts]: use companyId from request to get only opportunities for the needed company

File: backend/src/api/routes/opportunities-route.ts [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+);
+router.post(
+  '/',
+  run((req) => createOpportunities(req.body.opportunities)),
```
Comment (aleksandr-kuharenko) [inline backend/src/api/routes/opportunities-route.ts]: use companyId from request to create opportunities under the needed company
and userId as well

File: backend/src/data/seed-data/opp-tag.data.ts [hunk 18]
```diff
@@ -0,0 +1 @@
+export const opp_tags = [{}, {}, {}];
```
Comment (aleksandr-kuharenko) [inline backend/src/data/seed-data/opp-tag.data.ts]: it is not used anywhere

File: backend/src/services/opportunity.service.ts [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+export const getOpportunities = async (): Promise<Opportunity[]> => {
+  const opportunityRepository = getCustomRepository(OpportunitiesRepository);
+  const companyRepository = getCustomRepository(CompanyRepository);
+  const userRepository = getCustomRepository(UserRepository);
+  const companyInstance: Company = await companyRepository.findOne({
+    name: companies[0].name,
+  });
+  const userInstance: User = await userRepository.findOne({
+    firstName: users[0].firstName,
+  });
```
Comment (aleksandr-kuharenko) [inline backend/src/services/opportunity.service.ts]: use real companyId and userId from token

File: backend/src/services/opportunity.service.ts [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+): Promise<any> => {
+  const companyRepository = getCustomRepository(CompanyRepository);
+  const userRepository = getCustomRepository(UserRepository);
+  const opportunitiesRepository = getCustomRepository(OpportunitiesRepository);
+  const companyInstance: Company = await companyRepository.findOne({
+    name: companies[0].name,
+  });
+  const userInstance: User = await userRepository.findOne({
+    firstName: users[0].firstName,
+  });
```
Comment (aleksandr-kuharenko) [inline backend/src/services/opportunity.service.ts]: use real companyId and userId from token

**Review by aleksandr-kuharenko — CHANGES_REQUESTED** [review]

File: backend/src/services/opportunity.service.ts [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+    id: companyId,
+  });
+  const userInstance: User = await userRepository.findOne({
+    firstName: users[0].firstName,
+  });
+  const opportunities = await opportunityRepository.find({
+    company: companyInstance,
+    user: userInstance,
+    relations: ['company', 'user', 'tags'],
+  } as FindManyOptions);
```
Comment (aleksandr-kuharenko) [inline backend/src/services/opportunity.service.ts]: As I understood, this service should return all opportunities in the company. You do not need user here

**Review by aleksandr-kuharenko — CHANGES_REQUESTED** [review]

File: frontend/src/components/main-page/opportunities/opportunityList.tsx [hunk 22]
```diff
@@ -0,0 +1,34 @@
+import OpportunityItem from './opportunityItem';
+import './opportunities.scss';
```
Comment (nikita-remeslov) [inline frontend/src/components/main-page/opportunities/opportunityList.tsx]: rename to `styles.scss`

File: frontend/src/components/opportunities-page/opportunitiesPage.tsx [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+const OpprotunitiesPage: React.FC = () => {
+  const opportunities = useAppSelector(
+    (state) => state.opportunities.opportunities,
+  );
+  const dispatch = useAppDispatch();
+  const isLoaded = useAppSelector((state) => state.opportunities.isLoaded);
+  useEffect(() => {
+    isLoaded ? null : dispatch(opportunityActions.fetchLoadOpp());
+  }, []);
+  // const showModalHandler = ():void=> { dispatch(opportunitiesActions.showModal()); };
```
Comment (nikita-remeslov) [inline frontend/src/components/opportunities-page/opportunitiesPage.tsx]: Do you need these comments?

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]
Overall looks good

**Review by nikita-remeslov — APPROVED** [review]

**Review by aleksandr-kuharenko — APPROVED** [review]

**Review by AndriyCh1 — APPROVED** [review]

**Pull request merged**